uaclient/tests/test_data_types.py

```diff
@@ -0,0 +1,382 @@
+                {
+                    "string": "string",
+                    "integer": "1",
+                    "obj": {"string": "nestedstring", "integer": 2},
+                    "stringlist": ["one", "two"],
+                    "integerlist": [3, 4, 5],
+                    "objlist": [
+                        {"string": "nestedstring2", "integer": 6},
+                        {"string": "nestedstring3", "integer": 7},
+                    ],
+                },
+                IncorrectFieldTypeError(
+                    IncorrectTypeError("int", "1"), "integer"
+                ),
+            ),
+            (
+                {
+                    "string": "string",
+                    "integer": 1,
+                    "obj": {"string": 8, "integer": 2},
+                    "stringlist": ["one", "two"],
+                    "integerlist": [3, 4, 5],
+                    "objlist": [
+                        {"string": "nestedstring2", "integer": 6},
+                        {"string": "nestedstring3", "integer": 7},
+                    ],
+                },
+                IncorrectFieldTypeError(
+                    IncorrectFieldTypeError(
+                        IncorrectTypeError("string", 8), "string"
+                    ),
+                    "obj",
+                ),
+            ),
+            (
+                {
+                    "string": "string",
+                    "integer": 1,
+                    "obj": {"string": "nestedstring", "integer": 2},
+                    "stringlist": ["one", 2],
+                    "integerlist": [3, 4, 5],
+                    "objlist": [
+                        {"string": "nestedstring2", "integer": 6},
+                        {"string": "nestedstring3", "integer": 7},
+                    ],
+                },
+                IncorrectFieldTypeError(
+                    IncorrectListElementTypeError(
+                        IncorrectTypeError("string", 2), 1
+                    ),
+                    "stringlist",
+                ),
+            ),
+            (
+                {
+                    "string": "string",
+                    "integer": 1,
+                    "obj": {"string": "nestedstring", "integer": 2},
+                    "stringlist": ["one", "two"],
+                    "integerlist": [3, 4, 5],
+                    "objlist": [
+                        {"string": "nestedstring2", "integer": "6"},
+                        {"string": "nestedstring3", "integer": 7},
+                    ],
+                },
+                IncorrectFieldTypeError(
+                    IncorrectListElementTypeError(
+                        IncorrectFieldTypeError(
+                            IncorrectTypeError("int", "6"), "integer"
+                        ),
+                        0,
+                    ),
+                    "objlist",
+                ),
+            ),
+        ),
+    )
+    def test_error(self, val, error):
```

Comment:
Should we add a case where we don't pass a dict as a val ?

Reply:
Yes!